src/Core/src/Platform/iOS/KeyboardAutoManagerScroll.cs

```diff
@@ -0,0 +1,582 @@
+﻿/*
+ * This class is adapted from IQKeyboardManager which is an open-source
+ * library implemented for iOS to handle Keyboard interactions with
+ * UITextFields/UITextViews. Link to their MIT License can be found here:
+ * https://github.com/hackiftekhar/IQKeyboardManager/blob/7399efb730eea084571b45a1a9b36a3a3c54c44f/LICENSE.md
+ */
+
+using System;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using CoreGraphics;
+using Foundation;
+using UIKit;
+
+namespace Microsoft.Maui.Platform;
+
+internal static class KeyboardAutoManagerScroll
+{
+	static UIScrollView? LastScrollView;
+	static CGPoint StartingContentOffset;
+	static UIEdgeInsets StartingScrollIndicatorInsets;
+	static UIEdgeInsets StartingContentInsets;
+	static CGRect KeyboardFrame = CGRect.Empty;
+	static CGPoint TopViewBeginOrigin = new(nfloat.MaxValue, nfloat.MaxValue);
+	static readonly CGPoint InvalidPoint = new(nfloat.MaxValue, nfloat.MaxValue);
+	static double AnimationDuration = 0.25;
+	static UIView? View = null;
+	static UIView? ContainerView = null;
+	static CGRect? CursorRect = null;
+	internal static bool IsKeyboardShowing = false;
+	static int TextViewTopDistance = 20;
+	static int DebounceCount = 0;
+	static NSObject? WillShowToken = null;
+	static NSObject? WillHideToken = null;
+	static NSObject? TextFieldToken = null;
+	static NSObject? TextViewToken = null;
+
+	internal static void Init()
+	{
+		TextFieldToken = NSNotificationCenter.DefaultCenter.AddObserver(new NSString("UITextFieldTextDidBeginEditingNotification"), DidUITextBeginEditing);
+
+		TextViewToken = NSNotificationCenter.DefaultCenter.AddObserver(new NSString("UITextViewTextDidBeginEditingNotification"), DidUITextBeginEditing);
+
+		WillShowToken = NSNotificationCenter.DefaultCenter.AddObserver(new NSString("UIKeyboardWillShowNotification"), WillKeyboardShow);
+
+		WillHideToken = NSNotificationCenter.DefaultCenter.AddObserver(new NSString("UIKeyboardWillHideNotification"), WillHideKeyboard);
+	}
+
+	static async void DidUITextBeginEditing(NSNotification notification)
+	{
+		if (notification.Object is not null)
+		{
+			View = (UIView)notification.Object;
+
+			if (View is null)
+				return;
+
+			CursorRect = null;
+
+			ContainerView = View.SetContainerView();
+
+			// the cursor needs a small amount of time to update the position
+			await Task.Delay(5);
+
+			var localCursor = FindLocalCursorPosition();
+			if (localCursor is CGRect local)
+				CursorRect = View.ConvertRectToView(local, null);
+
+			TextViewTopDistance = localCursor is CGRect cGRect ? 20 + (int)cGRect.Height : 20;
```

Comment:
You are doing the same calculation twice when you do 'localCursor is CGRect'. Can you move to an as cast, and chante the if and ternary operators?
```suggestion
			var cGRect = localCursor as CGRect;
			if (cGRect is not null)
				CursorRect = View.ConvertRectToView(cGRect, null);

			TextViewTopDistance = (cGRect?.Heigh ?? 0) + 20;
```
Double check my logic in the top distance but I believe it is what you wanted to do.

Reply:
It says that you cannot use 'as' with a non-nullable value type such as CGRect, but I was able to do this to make the ternary operator not call 'is'

```cs
if (localCursor is CGRect local)
    CursorRect = View.ConvertRectToView(local, null);

TextViewTopDistance = ((int?)localCursor?.Height ?? 0) + 20;
```

Reply:
Oh! I forgot CGrect is a struct, not a ref , makes sense you do it like this then.